[WIP] Upgrade examples to terraform-google-vm and add test automation

zefdelgadillo opened the pull request:
- [x] Upgrade module to work with terraform-google-modules/terraform-google-vm
- [x] Upgrade simple example to use managed instance group in terraform-google-modules/terraform-google-vm
- [x] Replace existing test automation with new test automation using kitchen terraform
- [ ] Write a test to hit endpoint created
- [ ] Review README/documentation (having trouble getting `make generate_docs` to work correctly)

@morgante at long last, here's the (significant) PR that includes the change to use the new terraform-google-vm managed instance group examples. This included some refactor of the module in addition to adding the example. I also added the test automation that we have on the other newer modules, with a simple test for now that checks for the presence of the forwarding rule. We might want to include a test that checks for the presence of the backend.

morgante commented on examples/simple/main.tf:
```diff
@@ -0,0 +1,49 @@
+  region       = var.region
+  name         = var.network
+  service_port = 80
+  target_tags  = ["allow-lb-service"]
```
We should make this an output from the template. Also, I think we should encourage targeting by Service Account instead of tags.

zefdelgadillo replied:
Opened #18 to support Service Account in the module Firewall.

zefdelgadillo replied:
Opened terraform-google-modules/terraform-google-vm#33 for the output of the template module

morgante commented on examples/simple/variables.tf:
```diff
@@ -0,0 +1,41 @@
+
+variable region {
+}
+variable image_family {
```
Spacing please.

morgante commented on examples/simple/variables.tf:
```diff
@@ -0,0 +1,41 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+variable region {
```
We should set a default region.

morgante commented on examples/simple/variables.tf:
```diff
@@ -0,0 +1,41 @@
+  description = "project where source image comes from"
+  default     = "debian-cloud"
+}
+variable network {
```
The simplest possible example could (and should) build the network itself.

zefdelgadillo replied:
as in a `network.tf` file within the example directory that builds out a network/subnet/router/NAT?

morgante replied:
Right.

zefdelgadillo replied:
Moved as part of 136acf2, but ended up having to move the provider outside of the fixtures because of the way some of the regional resources are created. Not sure if this introduces too much in the example directory that might confuse users on how to implement the core load balancer.

morgante replied:
I don't quite follow. We shouldn't rely on the provider for location, instead we should explicitly set the region on all resources.

zefdelgadillo replied:
I've got this figured out. I was slightly misdirected by an error message pointing me to a missing region in the provider (in reality, the underlying issue was that I should have used `.self_link` instead of `.name` for a subnet reference).

morgante commented on examples/simple/variables.tf:
```diff
@@ -0,0 +1,41 @@
+variable subnetwork {
+  description = "subnetwork to provision managed instance group and load balancer"
+}
+variable "service_account" {
```
We should create the service account inline (part of the example) IMO.

zefdelgadillo replied:
+1 I think this makes this easier to reason about. I'll update

morgante commented on test/fixtures/shared/network.tf:
```diff
@@ -0,0 +1,40 @@
+# Copyright 2019 Google LLC
```
The "shared" pattern is confusing in many cases. I'd suggest we remove it since we only have a single example/fixture.

morgante commented on test/fixtures/shared/terraform.tfvars.example:
```diff
@@ -0,0 +1,6 @@
+project_id="your-project-id"
```
With the new testing approach, users shouldn't set tfvars directly—instead, the prepare stage prepares a project/environment.

morgante commented on test/fixtures/simple/variables.tf:
```diff
@@ -0,0 +1,31 @@
+  description = "The GCP region to deploy instances into"
+  default     = "us-east4"
+}
+variable "service_account" {
```
We should create this within the fixture.

morgante commented on test/setup/make_source.sh:
```diff
@@ -0,0 +1,28 @@
+# shellcheck disable=SC2086
+echo "export SERVICE_ACCOUNT_JSON='$(echo $sa_json | base64 --decode)'" >> ../source.sh
+
+client_email=$(echo "$sa_json" | base64 --decode | jq -r .client_email)
```
The SA which the VMs run as should *not* be the same SA used to create them.

morgante commented on test/task_helper_functions.sh:
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env bash
```
We shouldn't need a custom helper functions file for this module. Can we delete?

zefdelgadillo replied:
Removed

zefdelgadillo commented on examples/simple/main.tf:
```diff
@@ -0,0 +1,49 @@
+  vars = {
+    PROXY_PATH = ""
+  }
+}
```
I'll update to that standard, but `terraform fmt` doesn't seem to care about lines between resources.
```
zef:~/workspace/tmp/tf $ cat main.tf
resource "null_resource" "test" {

}
resource "null_resource" "test2" {

}
zef:~/workspace/tmp/tf $ terraform fmt .
zef:~/workspace/tmp/tf $ cat main.tf
resource "null_resource" "test" {

}
resource "null_resource" "test2" {

}
```

zefdelgadillo commented:
Requested changes I think should be resolved, and I opened a few additional issues where there are external dependencies. Added an additional test for presence of load balancer endpoint that gets created. Also updated README documentation to the same standards as some of the other CFT modules.

I'll touch base separately about where I am on adding CI.

zefdelgadillo commented:
The integration test I had to test the presence of the load balancer by hitting the endpoint fails for what it seems like the failure of the docker build container to access the internet. The basic functionality of the load balancer in GCP tests for the presence of the backend, but I think to complete an integration test we'd want to test the presence of the front end and forwarding rules.

Are we doing any tests for Internet-facing endpoints like this anywhere else?
```
control "http" do
  title "http"
    describe http("http://#{attribute("load_balancer_ip")}") do
      its('status') { should eq 200 }
  end
end
```
```
Step #3 - "verify":   Ã—  http: http
Step #3 - "verify":      Ã—  HTTP GET on http://34.70.209.72 status 
Step #3 - "verify":      execution expired
```

Jberlinsky commented:
@zefdelgadillo We had some full integration tests like what you're describing in other modules, but they've since (to my knowledge, entirely) been removed or commented out because they were "flappy" -- even when they worked, there was no way to guarantee they worked consistently.

Jberlinsky commented on test/integration/basic/controls/gcloud.rb:
```diff
@@ -0,0 +1,30 @@
+
+control "gcloud" do
+  title "gcloud"
+  describe command("gcloud compute forwarding-rules list --project=#{attribute("project_id")} --format json") do
```
Do we want to consider using [inspec-gcp](https://github.com/inspec/inspec-gcp/blob/master/docs/resources/google_compute_forwarding_rule.md) matchers here?

zefdelgadillo replied:
For the most part across other modules we've been using the local control and gcloud commands. I'm definitely open to using inspec-gcp because it does feel cleaner, but I'm always hesitant on introducing new dependencies to libraries - we can maintain the gcloud SDK version via our test docker image. @aaron-lane @morgante do you have an opinion on inspec-gcp vs. local gcloud tests?

morgante replied:
Either is fine. inspec-gcp is a bit cleaner, but really whichever is easiest/simplest to write should be preferred.

zefdelgadillo replied:
I’ll leave this for now but noted on the next test to consider using inspec-gcp. Appreciate the insight and discussion.

zefdelgadillo commented:
@morgante or @Jberlinsky thank you both again for the reviews. would like to get a final once over and an approval if all looks good

morgante commented on examples/basic/README.md:
```diff
@@ -1,80 +1,16 @@
+# Basic Example
+This example illustrates how to use the `lb` module with a managed instance group from [terraform-google-modules/terraform-google-vm].
+
+## Inputs
```
Needs to be autogenerated.

zefdelgadillo commented on examples/basic/README.md:
```diff
@@ -1,80 +1,16 @@
-<a href="https://concourse-tf.gcp.solutions/teams/main/pipelines/tf-examples-lb-basic" target="_blank">
-<img src="https://concourse-tf.gcp.solutions/api/v1/teams/main/pipelines/tf-examples-lb-basic/badge" /></a>
-
-This example creates a managed instance group with 2 instances in the same region and a network TCP Load Balancer.
```
I've re-added the explanation, diagram, and usage instructions with the link out to Cloud Shell to step through as a tutorial

morgante commented on CHANGELOG.md:
```diff
@@ -7,6 +7,16 @@ project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+## [2.1.0] - 2019-10-01
```
Technically no need for this to be a release.

zefdelgadillo replied:
Moved to unreleased.

morgante commented on README.md:
````diff
@@ -1,35 +1,55 @@
+Basic usage is as follows:
+```hcl
+module "load_balancer" {
+  source       = "terraform-google-modules/lb"
````
```suggestion
  source       = "terraform-google-modules/lb/google"
  version      = "~> 2.0.0"
```

zefdelgadillo replied:
Done

morgante commented on README.md:
```diff
@@ -1,35 +1,55 @@
-  service_port = "${module.mig1.service_port}"
-  target_tags  = ["${module.mig1.target_tags}"]
+module "managed_instance_group" {
+  source            = "terraform-google-modules/vm/google//modules/mig"
```
Please pin version.

zefdelgadillo replied:
Done

morgante commented on README.md:
````diff
@@ -1,35 +1,55 @@
 }
 ```
-
-## Resources created
````
Please preserve the resource and architecture diagrams.

zefdelgadillo replied:
Re-added
